### ISSUE TITLE
Installing hits 'C++11 standard requested but CXX11 is not defined'

### ISSUE THREAD

**Issue by javierluraschi** [rstudio/httpuv#82]
I'm hitting this in Amazon EMR, did the required compiler min-version for `httpuv` changed?

```
> install.packages("shiny")
Installing package into ‘/home/hadoop/R/x86_64-redhat-linux-gnu-library/3.4’
(as ‘lib’ is unspecified)
also installing the dependency ‘httpuv’

trying URL 'https://cran.rstudio.com/src/contrib/httpuv_1.3.5.tar.gz'
Content type 'application/x-gzip' length 427638 bytes (417 KB)
==================================================
downloaded 417 KB

trying URL 'https://cran.rstudio.com/src/contrib/shiny_1.0.5.tar.gz'
Content type 'application/x-gzip' length 2330710 bytes (2.2 MB)
==================================================
downloaded 2.2 MB

* installing *source* package ‘httpuv’ ...
** package ‘httpuv’ successfully unpacked and MD5 sums checked
** libs
Error in .shlib_internal(args) : 
  C++11 standard requested but CXX11 is not defined
* removing ‘/home/hadoop/R/x86_64-redhat-linux-gnu-library/3.4/httpuv’
Warning in install.packages :
  installation of package ‘httpuv’ had non-zero exit status
ERROR: dependency ‘httpuv’ is not available for package ‘shiny’
* removing ‘/home/hadoop/R/x86_64-redhat-linux-gnu-library/3.4/shiny’
Warning in install.packages :
  installation of package ‘shiny’ had non-zero exit status

The downloaded source packages are in
	‘/mnt/tmp/RtmpkvYP7J/downloaded_packages’
```

```
> system2("gcc", "--version")
gcc (GCC) 4.8.3 20140911 (Red Hat 4.8.3-9)
Copyright (C) 2013 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

```
> system2("cat", "/etc/*-release")
LSB_VERSION=base-4.0-amd64:base-4.0-noarch:core-4.0-amd64:core-4.0-noarch:printing-4.0-amd64:printing-4.0-noarch
NAME="Amazon Linux AMI"
VERSION="2017.03"
ID="amzn"
ID_LIKE="rhel fedora"
VERSION_ID="2017.03"
PRETTY_NAME="Amazon Linux AMI 2017.03"
ANSI_COLOR="0;33"
CPE_NAME="cpe:/o:amazon:linux:2017.03:ga"
HOME_URL="http://aws.amazon.com/amazon-linux-ami/"
Amazon Linux AMI release 2017.03
```

```
> sessionInfo()
R version 3.4.1 (2017-06-30)
Platform: x86_64-redhat-linux-gnu (64-bit)
Running under: Amazon Linux AMI 2017.03

Matrix products: default
BLAS/LAPACK: /usr/lib64/R/lib/libRblas.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_3.4.1 tools_3.4.1  
```

I'm almost sure the compiler version has not changed in EMR, especially not a downgrade, I'm also sure this installation used to work in EMR. I can upgrade the compiler myself, but this breaks some folks since there are scripts out there that won't support this, say: https://aws.amazon.com/blogs/big-data/running-sparklyr-rstudios-r-interface-to-spark-on-amazon-emr/. Therefore, would be good to know what has changed to properly upgrade scripts.

**Comment by javierluraschi** [comment]
Actually, the EMR script does use the latest R version... so they probably auto-upgraded customers using this script to `R 3.4.1`, taking a look...

**Comment by javierluraschi** [comment]
Confirmed, this is caused by upgrading to `R 3.4.1`, so not `httpuv` specific, sorry for the false alarm.

**Comment by dkincaid** [comment]
I am having the same problem. So how does one fix this problem? Doesn't httpuv work with R 3.4?

**Comment by javierluraschi** [comment]
Not sure this is really an `httpuv` issue but I also heard a customer hitting this expecting this to be smoother in `3.4`, at least, I can close with some more info for the community.

**Comment by jcheng5** [comment]
@dkincaid Are you also on EC2, and if so, what AMI are you using? Certainly httpuv doesn't have a problem with R 3.4 on most platforms/configurations so any help in identifying reproducible cases would be a big help.

**Comment by bfey** [comment]
I have also run into this issue just recently, using the script at [https://aws.amazon.com/blogs/big-data/running-sparklyr-rstudios-r-interface-to-spark-on-amazon-emr/](url)

my config:
release: emr-5.8.0
applications: Zeppelin 0.7.2, Ganglia 3.7.2, ZooKeeper 3.4.10, Spark 2.2.0
instance: c3.2xlarge
bootstrap actions: {"ScriptBootstrapAction":{"Path":"s3://aws-bigdata-blog/artifacts/aws-blog-emr-rstudio-sparklyr/rstudio_sparklyr_emr5.sh","Args":["--sparklyr","--rstudio","--rexamples","--plyrmr","--user","xxxxx","--user-pw","xxxxx"]},"Name":"R-install"}

I can confirm that it worked fine in July, and appears to be related to R 3.4.1. (even though latest_r is not passed to script in bootstrap). Making the changes to the Makeconf file described here: [https://github.com/Microsoft/microsoft-r-open/issues/26](url) allows me to get past the httpuv install error. But R appears to become unstable and use of sparklyr and dplyr results in bizarre errors afterward.

**Comment by dkincaid** [comment]
We are running into this using AWS EMR. I have tried versions 5.7.0 and 5.8.0 and both are doing this. For now, we have switched back to R version 3.3, but it would be really nice to get this figured out. Please let me know what other information I could provide or other things to try. Happy to help troubleshoot if needed.

**Comment by wch** [comment]
Does the EMR image use Microsoft R Open? According to this discussion, there's a bug in MRO: jeroen/protolite#5

And from Microsoft/microsoft-r-open#26, it sounds like it's present in their releases of 3.4.0 and 3.4.1.

**Comment by dkincaid** [comment]
No, it's running on Linux using the real R implementation.

**Comment by jcheng5** [comment]
@jeroen It looks like you added this line a few months ago: https://github.com/rstudio/httpuv/blob/65016f205bdfca17d7714a44740682272bd1c25c/src/Makevars#L4

Removing this line fixes it. Are there any drawbacks to doing so?

**Comment by jeroen** [comment]
If I recall correctly there was some compiler warning about a C++11 feature in your code.

**Comment by jeroen** [comment]
The `CXX11` variable not being defined is a bug in the `EMR` R configuration. The majority of packages now uses `CXX_STD=CXX11` so removing this in httpuv is not really solving anything.

**Comment by javierluraschi** [comment]
@jeroen are you recommending to define an `CXX11` environment variable and reinstall `httpuv`? Or is this rather a variable that should have been set in EMR while they built their R binaries?

I can try the former option right away, what exactly is the value we would need to set this to?

**Comment by jeroen** [comment]
This C++ variables must be set when R is compiled. They get stored in this file:

```
cat $(R RHOME)/etc/Makeconf
```

You can fix it by manually editing the file to set the correct variables. The values below are from the standard R from EPEL on CentOS 7:

```make
CXX = g++ -m64
CXXCPP = $(CXX) -E
CXXFLAGS = -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic $(LTO)
CXXPICFLAGS = -fpic
CXX98 = g++ -m64
CXX98FLAGS = -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic
CXX98PICFLAGS = -fpic
CXX98STD =
CXX11 = g++ -m64
CXX11FLAGS = -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic
CXX11PICFLAGS = -fpic
CXX11STD = -std=gnu++11
```

However the real solution is that the maintainer of the R package in emr sets this correctly.

**Comment by wch** [comment]
This is what I get from the EMR image:

```
$ R CMD config CXX11

$ R CMD config CXX11FLAGS

$ rpm -qf /usr/lib64/R/etc/Makeconf
R-core-3.4.1-1.52.amzn1.x86_64
```

Poking around the system, I noticed that in /usr/lib64/R/etc, there's a Makeconf and a Makeconf.rpmnew file. I copied the .rpmnew file over the original, and after doing that, I get:

```
$ R CMD config CXX11
g++ -m64
$ R CMD config CXX11FLAGS
-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic
```

Also, I was able to install httpuv without any trouble.

The .rpmnew file is put there because the system is cautious when upgrading RPM packages. So I suspect that what happened is this: The creator of the AMI installed an older version of R with this problem. Then, when they upgraded R, the new Makeconf file was saved as Makeconf.rpmnew instead of replacing the original.

Apparently the packager can decide whether the new files replace the old ones, so whoever maintains the RPM package should probably make it replace Makeconf. See here for more:
  https://serverfault.com/questions/48776/why-do-i-have-rpmnew-file-after-an-update

**Comment by javierluraschi** [comment]
@jcheng5 @wch the suggestion to copy as in:

```
sudo cp /usr/lib64/R/etc/Makeconf.rpmnew /usr/lib64/R/etc/Makeconf
```

works here, just verified this on a new cluster. I'll pass that to our customers and find/follow-up with the AMI maintainer, somehow, to get this fixed in EMR.

Feel free to close this one, thank you!

**Comment by jeroen** [comment]
Thanks for investigating this @wch! Can somebody notify the maintainer of the rpm? I've seen this issue reported on several different packages.

**Comment by javierluraschi** [comment]
@jeroen yes, I will.

**Comment by dkincaid** [comment]
Awesome work, guys! Thank you all so much for troubleshooting this even though it isn't technically an httpuv problem. Very much appreciate it!

**Comment by wch** [comment]
I think I was slightly incorrect in my description of how the .rpmnew file came to exist on the AMI. A .rpmnew file is created only if the file has been changed since it was originally installed. So the AMI creator must have installed a previous version of R, edited the Makeconf file, and then installed the new version.

Knowing this, I feel less sure that the behavior of the RPM package should be changed. (Of course, the AMI still needs to be fixed.)

**Comment by jeroen** [comment]
Yes that is correct. For completeness, this only happens for files under `/etc`, i.e. config files. 

When the package manager (`rpm`/`dpkg`) installs an update of a package; it first checks if files under `/etc/` have been modified by the user. Unchanged files are automatically replaced with the new version. For files that have been changed, you usually get a prompt what you want to do (keep the old file or the new). Perhaps in non-interactive mode it will just keep the old one by default.

I'm surprised though that an upgrade is involved when the AMI is created. Usually the AMI can just install the latest version directly.

**Comment by tomz** [comment]
The EMR SparklyR bootstrap action (https://aws.amazon.com/blogs/big-data/running-sparklyr-rstudios-r-interface-to-spark-on-amazon-emr/) is fixed to work with R 3.4.1 now

**Comment by luckylara** [comment]
sudo R CMD INSTALL rstan_2.17.4.tar.gz 
[sudo] password for lara: 
* installing to library ‘/home/lara/R/x86_64-pc-linux-gnu-library/3.4’
* installing *source* package ‘rstan’ ...
** package ‘rstan’ successfully unpacked and MD5 sums checked
** libs
Error in .shlib_internal(args) : 
  C++14 standard requested but CXX14 is not defined
* removing ‘/home/lara/R/x86_64-pc-linux-gnu-library/3.4/rstan’
``
